Updates 'grovetccfg' to to write a grove.cfg from the grove servers profile

jrushford opened the pull request:
Modified grovetccfg to read a grove server profile from the traffic_ops and write a grove.cfg from that profile id there have been changes. Fixes #2607 and depends on #2517 and #2523. Both #2517 and #2523 are merged onto master.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2161/
Test FAILed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2165/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2166/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2167/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2170/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2171/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2172/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2174/
Test PASSed.

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -259,20 +328,162 @@ func main() {
+func createGroveCfg(toc *to.Session, server tc.Server) (bool, config.Config, error) {
+	var new_cfg config.Config
+	var cur_cfg config.Config
+	var plugin_params = []string{}
```
Go local variables should be `camelCase` not `snake_case`

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -259,20 +328,162 @@ func main() {
+
+func setConfigParameter(cfg *config.Config, name string, value string) error {
+	var err error
+	err = nil
```
This is unnecessary, Go zero-initializes all variables, so `err` is already `nil`.

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -259,20 +328,162 @@ func main() {
+func setConfigParameter(cfg *config.Config, name string, value string) error {
+	var err error
+	err = nil
+	if name == "rfc_compliant" {
```
A `switch` would be cleaner here.
```
switch name {
case "rfc_compliant":
  cfg.RFCCompliant, err = strconv.ParseBool(value)
case "remap_rules_file":
  cfg.RemapRulesFile = value
...
}
return err
```

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -94,13 +99,13 @@ func CopyAndGzipFile(src, dst string) error {
 
 // BackupFile copies the given file to a new file in the subdirectory "/remap_history", with the name suffixed by the current timestamp. Returns nil if the given file doesn't exist (nothing to back up).
-func BackupFile(path string) error {
+func BackupFile(path string, bkup_dir string) error {
```
`bkup_dir` should be camel case, and while Go favors brevity `bkup` is a bit extreme, I'd recommend `backupDir`

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -202,11 +207,75 @@ func main() {
+	var hostProfile tc.Profile
+	var ok bool
+	var profiles map[string]tc.Profile
+	var servers map[string]tc.Server
```
If `host == nil`, this gets passed to `createRulesOldAPI` which will panic. This should check if `host == nil`, and print a good error message and `os.Exit(1)`.

That will also reduce indentation, and reduce variable pre-declarations too, which will make this block much nicer.

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -202,11 +207,75 @@ func main() {
+	} // end of API 1.2 stuff
+
+	if hostProfile.Type == GroveProfileType {
+		update_required, cfg, err := createGroveCfg(toc, hostServer)
```
`update_required` should be `updateRequired`

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -259,20 +328,162 @@ func main() {
+	} else {
+		// load config parameters from the servers profile
+		for _, p := range serverParameters {
+			if p.ConfigFile == "grove.cfg" {
```
The config file is duplicated in `GroveConfigPath`. Can you change the variables to 
```
const GroveConfigFile = "grove.cfg"
const GroveConfigPath = "/etc/grove/" + GroveConfigFile
```

To get rid of the magic string?

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -259,20 +328,162 @@ func main() {
+
+	// load the servers current config parameters.
+	if _, err := os.Stat(GroveConfigPath); err == nil {
+		cur_cfg, err = config.LoadConfig(GroveConfigPath)
```
`cur_cfg` should be `currentCfg` or `currCfg`

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -259,20 +328,162 @@ func main() {
+		cfg.RFCCompliant, err = strconv.ParseBool(value)
+		return err
+	}
+	if name == "port" {
```
All these strings duplicating the JSON tags aren't great, but unfortunately, I don't think Go has a better way to do this

rob05c commented on grove/grovetccfg/grovetccfg.go:
```diff
@@ -50,6 +52,9 @@ const UserAgent = "grove-tc-cfg/" + Version
 const GroveConfigPath = "/etc/grove/grove.cfg"
+const ConfigHistory = "cfg_history/"
+const RemapHistory = "remap_history/"
+const GroveProfileType = "GROVE_PROFILE"
```
Would you mind moving this to `traffic_control/lib/go-tc/enum.go`? Profile types should be there, so other apps like the TO client can use them.

rob05c commented:
Ok, I commented on all the formatting things I saw. I didn't do an actual review, because that will take longer, and I'll need to install and run it against a TO instance with the expected profile. I'll try to do that as soon as I can find the time.

Feel free to fix the things noted now, or wait until the full review, whatever you prefer.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2189/
Test FAILed.

jrushford commented:
Hey Rob,

I’ve made the requested changes except for this one, I do not see that there is currently any profile enum in go-tc.  If necessary, I could add one I suppose.
I had several commits so, I’ve also squashed them.

If it helps, on ipcdn-cache-59.cdnlab.comcast.net, I have a cdn-in-a-box running with the current database from staging where I have added the GROVE_PROFILE to profile_type_values and  have a EDGE1_XCR_GROVE_0.1-9001 built.  I’ve been running this script on cache-59 for testing, you could use my hostname below or pick another and change it’s profile to use this one:

#!/bin/bash
./grovetccfg -ignore-update-flag -host cdn-ec-sfb-001-01 -touser jrushford -topass twelve -insecure -tourl https://localhost:6443 -pretty

From: Robert Butts <notifications@github.com>
Reply-To: apache/trafficcontrol <reply@reply.github.com>
Date: Monday, August 6, 2018 at 9:04 AM
To: apache/trafficcontrol <trafficcontrol@noreply.github.com>
Cc: John Rushford <John_Rushford@cable.comcast.com>, Author <author@noreply.github.com>
Subject: [EXTERNAL] Re: [apache/trafficcontrol] Updates 'grovetccfg' to to write a grove.cfg from the grove servers profile (#2606)


Ok, I commented on all the formatting things I saw. I didn't do an actual review, because that will take longer, and I'll need to install and run it against a TO instance with the expected profile. I'll try to do that as soon as I can find the time.

Feel free to fix the things noted now, or wait until the full review, whatever you prefer.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/apache/trafficcontrol/pull/2606#issuecomment-410739702>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AEVyblOt5KJLSwGGGOEsofkJihPmO-ONks5uOFrogaJpZM4VrE7r>.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2191/
Test FAILed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2194/
Test PASSed.

jrushford commented:
@dewrich, @dangogh, @DylanVolz can i get a review of this pr.  It is pretty straight forward.  It writes a grove.cfg file for a server when a grove profile is used.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/trafficcontrol-PR/2352/
Test PASSed.
